src/whisper.cpp

```diff
@@ -4343,243 +4356,1383 @@ const char * whisper_print_system_info(void) {
+    std::vector<ggml_context *> ctxs;
 
-    WHISPER_ASSERT(is_positive_char || pos->type == WHISPER_GRETYPE_CHAR_NOT); // NOLINT
+    // buffer for the model tensors
+    std::vector<ggml_backend_buffer_t> buffers;
 
-    do {
-        if (pos[1].type == WHISPER_GRETYPE_CHAR_RNG_UPPER) {
-            // inclusive range, e.g. [a-z]
-            found = found || (pos->value <= chr && chr <= pos[1].value);
-            pos += 2;
-        } else {
-            // exact char match, e.g. [a] or "a"
-            found = found || pos->value == chr;
-            pos += 1;
-        }
-    } while (pos->type == WHISPER_GRETYPE_CHAR_ALT);
+    // tensors
+    int n_loaded;
+    std::map<std::string, struct ggml_tensor *> tensors;
+};
 
-    return std::make_pair(found == is_positive_char, pos);
-}
+struct whisper_vad_state {
+    int64_t t_vad_us = 0;
+    std::vector<ggml_backend_t> backends;
 
-// returns true iff some continuation of the given partial UTF-8 sequence could satisfy the char
-// range at pos (regular or inverse range)
-// asserts that pos is pointing to a char range element
-static bool whisper_grammar_match_partial_char(
-        const whisper_grammar_element * pos,
-        const whisper_partial_utf8      partial_utf8) {
+    struct ggml_tensor * h_state;
+    struct ggml_tensor * c_state;
 
-    bool is_positive_char = pos->type == WHISPER_GRETYPE_CHAR;
-    WHISPER_ASSERT(is_positive_char || pos->type == WHISPER_GRETYPE_CHAR_NOT);
+    ggml_backend_buffer_t buffer = nullptr;
 
-    uint32_t partial_value = partial_utf8.value;
-    int      n_remain      = partial_utf8.n_remain;
+    std::vector<uint8_t> ctx_buf;
 
-    // invalid sequence or 7-bit char split across 2 bytes (overlong)
-    if (n_remain < 0 || (n_remain == 1 && partial_value < 2)) {
-        return false;
-    }
+    whisper_sched sched;
 
-    // range of possible code points this partial UTF-8 sequence could complete to
-    uint32_t low  = partial_value << (n_remain * 6);
-    uint32_t high = low | ((1 << (n_remain * 6)) - 1);
+    std::vector<float> probs;
+};
 
-    if (low == 0) {
-        if (n_remain == 2) {
-            low = 1 << 11;
-        } else if (n_remain == 3) {
-            low = 1 << 16;
-        }
-    }
+struct whisper_vad_segment {
+    float start; // Start time in seconds
+    float end;   // End time in seconds
+};
 
-    do {
-        if (pos[1].type == WHISPER_GRETYPE_CHAR_RNG_UPPER) {
-            // inclusive range, e.g. [a-z]
-            if (pos->value <= high && low <= pos[1].value) {
-                return is_positive_char;
-            }
-            pos += 2;
-        } else {
-            // exact char match, e.g. [a] or "a"
-            if (low <= pos->value && pos->value <= high) {
-                return is_positive_char;
-            }
-            pos += 1;
-        }
-    } while (pos->type == WHISPER_GRETYPE_CHAR_ALT);
+struct whisper_vad_timestamps {
+    int n_segments;
+    std::vector<whisper_vad_segment> segments;
+};
 
```

Comment:
Can simplify now by removing `n_segments` member:

```suggestion
struct whisper_vad_timestamps {
    std::vector<whisper_vad_segment> segments;
};

```

A better name for `struct whisper_vad_timestamps` now might be `struct whisper_vad_segments`